lib/Twig/Template.php

```diff
@@ -555,37 +555,48 @@ protected function getAttribute($object, $item, array $arguments = array(), $typ
         $class = get_class($object);
 
         // object method
-        if (!isset(self::$cache[$class]['methods'])) {
+        if (null === $methodsCache = &self::$cache[$class]['methods']) {
             // get_class_methods returns all methods accessible in the scope, but we only want public ones to be accessible in templates
             if ($object instanceof self) {
                 $ref = new ReflectionClass($class);
                 $methods = array();
 
                 foreach ($ref->getMethods(ReflectionMethod::IS_PUBLIC) as $refMethod) {
-                    $methodName = strtolower($refMethod->name);
-
                     // Accessing the environment from templates is forbidden to prevent untrusted changes to the environment
-                    if ('getenvironment' !== $methodName) {
-                        $methods[$methodName] = true;
+                    if ('getenvironment' !== strtolower($refMethod->name)) {
+                        $methods[] = $refMethod->name;
                     }
                 }
-
-                self::$cache[$class]['methods'] = $methods;
             } else {
-                self::$cache[$class]['methods'] = array_change_key_case(array_flip(get_class_methods($object)));
+                $methods = get_class_methods($object);
+            }
+            $methodsCache = array();
+
+            foreach ($methods as $method) {
+                $methodsCache[$method] = $method;
+                $methodsCache[$name = strtolower($method)] = $method;
+
+                if (0 === strpos($name, 'get')) {
+                    $name = substr($name, 3);
+                } elseif (0 === strpos($name, 'is')) {
+                    $name = substr($name, 2);
+                } else {
+                    continue;
+                }
+
+                if (!isset($methodsCache[$name])) {
+                    $methodsCache[$name] = $method;
+                }
             }
         }
 
         $call = false;
-        $lcItem = strtolower($item);
-        if (isset(self::$cache[$class]['methods'][$lcItem])) {
-            $method = (string) $item;
-        } elseif (isset(self::$cache[$class]['methods']['get'.$lcItem])) {
-            $method = 'get'.$item;
```

Comment:
optim 3: zero string copy vs concat or cast to string (which is free for strings on PHP7, but copies on PHP5)